### PR TITLE
[wallet] Restore ability to list incoming transactions by label

### LINKED ISSUE #14382: Label support seems insufficient to replace accounts
While most of the additional functionality the old account system had over the label system were irrelevant, there is one critical functionality that no longer seems to be possible with labels - specifically, the ability to request a list of transactions (TXID + amount) that pay to addresses with a given account/label. As per the release notes for 0.17.0. the _listtransaction_ method was changed as follows;

> The account named parameter has been renamed to dummy. If provided, the dummy parameter must be set to the string *, unless running with the -deprecatedrpc=accounts argument (in which case functionality is unchanged).

_getreceivedbylabel_ and _getreceivedbyaddress_ only returns the balance, while _listreceivedbylabel_ and _listreceivedbyaddress_ do not actually filter; additionally, the former does not list transactions at all, while the latter does not include the amounts per TXID. While you could use _getaddressesbylabel_ to iterate over addresses, there is no corresponding method to get the TXIDs and amounts from this list of addresses. None of these can therefore be used as a direct replacement for this method.

The only way I can see to get anything resembling the old functionality with the API changes is to dump the entire transaction history with _listreceivedbyaddress_, filter the result by label manually, then use _gettransaction_ on each listed TXID. This is both very fiddly and very inefficient, and likely impossible when you go past a certain number of lifetime transactions.

Am I missing something here, or will it really no longer possible be to iterate over transactions received by account/label? If so, this basically means that many applications will no longer be usable with releases starting from Bitcoin Core 0.18.

### PR DESCRIPTION
**Pull request by ryanofsky**
This change partially reverts #13075 and #14023.

Fixes #14382

### CODE REVIEW

**Comment by MarcoFalke** [comment]
Is this for backport to 0.17.1? While not necessary, I believe it would simplify upgrade from 0.17.1+ to 0.18.0, because 0.17.1 could be used without `-deprecatedrpc` as workaround.

**Comment by ryanofsky** [comment]
I think it'd be reasonable to backport this, since it's a small change that restores a removed feature. But I think it won't cherry-pick cleanly because #14023 was merged since the 0.17 branch. @Saicere originally reported the issue so they might be able to say whether there's a benefit to backporting.

**Comment by Saicere** [comment]
> @Saicere originally reported the issue so they might be able to say whether there's a benefit to backporting.

As far as I am concerned, it's not really a big deal whether or not it's backported to the 0.17 branch as long as it's in before accounts are removed entirely. If you were using accounts in your application, you are probably running 0.17 with `-deprecatedrpc=accounts` anyway while RPC-facing code is updated.

**Comment by kristapsk** [comment]
ACK 8fcb76508493cd09d4895e7f3ad11cc5b945056e

File: src/wallet/rpcwallet.cpp [hunk 1]
```diff
@@ -1375,8 +1375,9 @@ static void MaybePushAddress(UniValue & entry, const CTxDestination &dest)
  * @param  fLong      Whether to include the JSON version of the transaction.
  * @param  ret        The UniValue into which the result is stored.
  * @param  filter     The "is mine" filter bool.
+ * @param  filter_label Optional label string to filter incoming transactions.
  */
-static void ListTransactions(CWallet* const pwallet, const CWalletTx& wtx, int nMinDepth, bool fLong, UniValue& ret, const isminefilter& filter) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
+static void ListTransactions(CWallet* const pwallet, const CWalletTx& wtx, int nMinDepth, bool fLong, UniValue& ret, const isminefilter& filter, const std::string* filter_label) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
```
Comment (promag) [inline src/wallet/rpcwallet.cpp]: >  @param  filter_label Optional label string to filter incoming transactions.

nit, `filter_label = nullptr`. Also, makes the diff smaller

**Review by promag — COMMENTED** [review]
Concept ACK.

**Comment by jnewbery** [comment]
Concept ACK. For completeness, this functionality was hidden behind a deprecation switch in #12953 (based on #11497). The functionality was marked deprecated in #5575.

Apologies - it was my oversight that this functionality got dropped. I'm happy to backport this to V0.17 once this gets merged. I aim to review this week.

**Review by jnewbery — COMMENTED** [review]
Tested ACK 8fcb76508493cd09d4895e7f3ad11cc5b945056e with a couple of nits.

@MarcoFalke - I'll go ahead and open a PR for backporting. A couple of questions:
- does the backport need release notes?
- if we add release notes for the backport, do we also need release notes here?

**Comment by jnewbery** [comment]
Backport is #14441. PR numbering was totally intentional.

**Review by ryanofsky — COMMENTED** [review]
Added 1 commit 8fcb76508493cd09d4895e7f3ad11cc5b945056e -> 665d00fce4bc53a9ea9dc5a7d457d586af357bf3 ([pr/list-label.1](https://github.com/ryanofsky/bitcoin/commits/pr/list-label.1) -> [pr/list-label.2](https://github.com/ryanofsky/bitcoin/commits/pr/list-label.2), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/list-label.1...pr/list-label.2)) with suggested fixes.
Squashed 665d00fce4bc53a9ea9dc5a7d457d586af357bf3 -> fa7fae5c3612cf77c91b92a39c14e73478424e85 ([pr/list-label.2](https://github.com/ryanofsky/bitcoin/commits/pr/list-label.2) -> [pr/list-label.3](https://github.com/ryanofsky/bitcoin/commits/pr/list-label.3))
Updated fa7fae5c3612cf77c91b92a39c14e73478424e85 -> 4deba4ced9fc4edd1a40837c083080b6ed4c8163 ([pr/list-label.3](https://github.com/ryanofsky/bitcoin/commits/pr/list-label.3) -> [pr/list-label.4](https://github.com/ryanofsky/bitcoin/commits/pr/list-label.4)) with release notes changes to reflect the intent to backport this to 0.17.1

File: src/wallet/rpcwallet.cpp (hunk 1 above)
Comment (jnewbery) [reply src/wallet/rpcwallet.cpp]: I agree that default arguments should be used sparingly.

**Comment by jnewbery** [comment]
The release note should be removed from this PR (assuming #14441 is merged)

File: src/wallet/rpcwallet.cpp (hunk 1 above)
Comment (promag) [reply src/wallet/rpcwallet.cpp]: :+1:

File: src/wallet/rpcwallet.cpp [hunk 2]
```diff
@@ -1460,10 +1464,12 @@ UniValue listtransactions(const JSONRPCRequest& request)
 
     if (request.fHelp || request.params.size() > 4)
         throw std::runtime_error(
-            "listtransactions (dummy count skip include_watchonly)\n"
+            "listtransactions (label count skip include_watchonly)\n"
```
Comment (promag) [inline src/wallet/rpcwallet.cpp]: nit, fix consistency by adding spaces after `(` and before `)`?

**Comment by ryanofsky** [comment]
Updated 4deba4ced9fc4edd1a40837c083080b6ed4c8163 -> 7cbe74f1524fc8621eb339d5eddd9530826c8461 ([pr/list-label.4](https://github.com/ryanofsky/bitcoin/commits/pr/list-label.4) -> [pr/list-label.5](https://github.com/ryanofsky/bitcoin/commits/pr/list-label.5)) removing release notes and changing whitespace as suggested.

**Comment by sipa** [comment]
Very nitty comment; it feels like this functionality is closer to `listreceivedbylabel` than to `listtransactions` (as the latter is about all transactions which affect the balance of the wallet - and formerly account) rather than just incoming payments. Would it make sense to add it as an extra option to `listreceivedbylabel` (verbose or so), or a separate RPC, instead?

**Comment by Saicere** [comment]
> Would it make sense to add it as an extra option to `listreceivedbylabel` (verbose or so), or a separate RPC, instead?

`listreceivedbylabel` doesn't take a label argument or list transactions in any way, it just lists total received coins per label, as per [the code](https://github.com/bitcoin/bitcoin/blob/be992701b018f256db6d64786624be4cb60d8975/src/wallet/rpcwallet.cpp#L1318).

**Comment by jnewbery** [comment]
ACK 7cbe74f1524fc8621eb339d5eddd9530826c8461. Only change since 4deba4c is removing release notes and changing whitespace.

**Comment by ryanofsky** [comment]
> Would it make sense to add it as an extra option to listreceivedbylabel (verbose or so), or a separate RPC, instead?

It'd be good to improve listreceivedbylabel. But filtering by label in `listtransactions` used to work previously and I think was just removed by accident.

You could maybe say that `listtransactions` shouldn't return labels or filter by label at all, since outgoing transactions won't have labels. But I don't think it makes sense to keep returning labels and only remove the ability to filter them.

File: src/wallet/rpcwallet.cpp [hunk 3]
```diff
@@ -1375,8 +1375,9 @@ static void MaybePushAddress(UniValue & entry, const CTxDestination &dest)
  * @param  fLong      Whether to include the JSON version of the transaction.
  * @param  ret        The UniValue into which the result is stored.
  * @param  filter     The "is mine" filter bool.
```
Comment (MeshCollider) [inline src/wallet/rpcwallet.cpp]: IMO this should be renamed to avoid confusion, to something like `filter_ismine`

**Review by meshcollider — COMMENTED** [review]
utACK https://github.com/bitcoin/bitcoin/pull/14411/commits/7cbe74f1524fc8621eb339d5eddd9530826c8461

**Comment by ryanofsky** [comment]
Rebased 7cbe74f1524fc8621eb339d5eddd9530826c8461 -> 671d8ee370de4824f4ab30b904bf59514f5e52e5 ([pr/list-label.5](https://github.com/ryanofsky/bitcoin/commits/pr/list-label.5) -> [pr/list-label.6](https://github.com/ryanofsky/bitcoin/commits/pr/list-label.6)) due to conflict with #14437. Also renamed filter variable as suggested by MeshCollider.

**Comment by MarcoFalke** [comment]
Unless I am mistaken this has been "backported" in 5150accdd2a7c7f0edf964d56bd7d34b5f740cdc. So this must go into 0.18.0 instead.
(assigned new milestone)

File: src/wallet/rpcwallet.cpp [hunk 4]
```diff
@@ -1352,10 +1356,12 @@ UniValue listtransactions(const JSONRPCRequest& request)
 
     if (request.fHelp || request.params.size() > 4)
         throw std::runtime_error(
-            "listtransactions (dummy count skip include_watchonly)\n"
+            "listtransactions ( label count skip include_watchonly )\n"
```
Comment (MarcoFalke) [inline src/wallet/rpcwallet.cpp]: From the documentation below, label is a string, so should be surrounded by quotation marks in the oneline doc. 

```suggestion
            "listtransactions ( \"label\" count skip include_watchonly )\n"
```

**Review by ryanofsky — COMMENTED** [review]
Rebased 671d8ee370de4824f4ab30b904bf59514f5e52e5 -> da427dbd48ae8d12a2a79a7514a813e78064fe52 ([pr/list-label.6](https://github.com/ryanofsky/bitcoin/commits/pr/list-label.6) -> [pr/list-label.7](https://github.com/ryanofsky/bitcoin/commits/pr/list-label.7)) due to conflict with #14720

**Comment by MarcoFalke** [comment]
ACK da427dbd48ae8d12a2a79a7514a813e78064fe52 (Only checked that the rebase was done properly from the last time people reviewed this. Also checked that the tests are the same as in the backport 5150accdd2a7c7f0edf964d56bd7d34b5f740cdc, didn't look at the code)

**Comment by jnewbery** [comment]
utACK da427dbd48ae8d12a2a79a7514a813e78064fe52

**Pull request merged**